crates/rune/src/diagnostics.rs

```diff
@@ -216,13 +220,90 @@ impl EmitDiagnostics for VmError {
+            }
+            VmErrorKind::UnsupportedBinaryOperation { lhs, rhs, op } => {
+                labels.push(
+                    Label::primary(source_id, span.range())
+                        .with_message("in this expression".to_string()),
+                );
+
+                (
+                    format!("type mismatch for operation `{}`", op),
+                    vec![
+                        format!("left hand side has type `{}`", lhs),
+                        format!("rhs hand side has type `{}`", rhs),
+                    ],
+                )
+            }
+            VmErrorKind::BadArgumentCount { actual, expected } => {
+                labels.push(
+                    Label::primary(source_id, span.range())
+                        .with_message("in this function call".to_string()),
+                );
 
+                (
+                    format!("wrong number of arguments"),
+                    vec![
+                        format!("expected `{}`", expected),
+                        format!("got `{}`", actual),
+                    ],
+                )
+            }
+            e => {
+                labels.push(
+                    Label::primary(source_id, span.range())
+                        .with_message("in this expression".to_string()),
+                );
+                ("internal vm error".to_owned(), vec![e.to_string()])
+            }
+        };
+
+        for ip in frames.iter().map(|v| v.ip()) {
+            let debug_inst = match debug_info.instruction_at(ip) {
+                Some(debug_inst) => debug_inst,
+                None => {
+                    writeln!(
+                        out,
+                        "virtual machine error: {} (no debug instruction)",
+                        error
+                    )?;
+
+                    return Ok(());
+                }
+            };
+
+            let source_id = debug_inst.source_id;
+            let span = debug_inst.span;
+
+            backtrace.push(StackFrame { source_id, span });
+        }
         let diagnostic = Diagnostic::error()
-            .with_message("virtual machine error")
-            .with_labels(labels);
+            .with_message(reason)
+            .with_labels(labels)
+            .with_notes(notes);
 
         term::emit(out, &config, &files, &diagnostic)?;
+
+        writeln!(out, "Callstack:")?;
+        for frame in &backtrace {
+            use codespan_reporting::files::Files;
+
+            let line = files
+                .line_index(frame.source_id, frame.span.start.into_usize())
+                .unwrap();
+            let line = files.line_number(frame.source_id, line).unwrap() - 1;
+            let line_range = files.line_range(frame.source_id, line).expect("a range");
+            let source = files.get(frame.source_id)?;
+            let name = source.name();
+            let slice = &source.source()[line_range];
```

Comment:
cleanup this with a helper function

Reply:
Maybe this should take the span from the inst instead... This is a bit odd right now with closures, you can end up repeating the same multiple times.